btc/utils.go

```diff
@@ -86,6 +177,87 @@ func getNetworkFromString(network string) (*chaincfg.Params, error) {
 	case "testnet":
 		return &chaincfg.TestNet3Params, nil
 	default:
-			return nil, errors.New("Invalid network")
+		return nil, errors.New("Invalid network")
+	}
+}
+
+// returns master private key from seed
+func getMasterKey(seed []byte, network string) (*hdkeychain.ExtendedKey, error) {
+	net, err := getNetworkFromString(network)
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := hdkeychain.NewMaster(seed, net)
+	if err != nil {
+		return nil, err
+	}
+
+	return key, nil
+}
+
+func newRedeemScript(m int, n int, pubkeys []string) (string, error) {
+	// check if params are valid:
+	// 1 <= n <= 7 && 1 <= m <= n
+	if n < 1 || n > 7 {
+		return "", errors.New("Invalid N param: it must be a value between 1 and 7 (inclusive)")
+	}
+	if m < 1 || m > n {
+		return "", errors.New("Invalid M param: it must be between 1 and N (inclusive)")
+	}
+
+	if len(pubkeys) != n {
+		return "", errors.New("Invalid number of pub keys: provided " + string(len(pubkeys)) + " expected " + string(n))
 	}
-}
\ No newline at end of file
+
+	// get OP Code for m and n
+	mOPCode := txscript.OP_1 + (m - 1)
+	nOPCode := txscript.OP_1 + (n - 1)
+	// multisig redeemScript format:
+	// <OP_m> <A pubkey> <B pubkey> <C pubkey>... <OP_n> OP_CHECKMULTISIG
+	var redeemScript bytes.Buffer
+	redeemScript.WriteByte(byte(mOPCode))
+	for _, pubkey := range pubkeys {
+		pubkeybytes, err := hex.DecodeString(pubkey)
+		if err != nil {
+			return "", err
+		}
+		redeemScript.WriteByte(byte(len(pubkeybytes))) //PUSH
+		redeemScript.Write(pubkeybytes)                //<pubkey>
+	}
+	redeemScript.WriteByte(byte(nOPCode))
+	redeemScript.WriteByte(byte(txscript.OP_CHECKMULTISIG))
+
+	return hex.EncodeToString(redeemScript.Bytes()), nil
+}
+
+// returns p2sh address derived from redeem script of multisig
+func getMultiSigAddress(redeemScript string, network string) (string, error) {
+	net, err := getNetworkFromString(network)
+	if err != nil {
+		return "", err
+	}
+
+	redeemScriptBytes, err := hex.DecodeString(redeemScript)
+	if err != nil {
+		return "", err
+	}
+
+	address, err := btcutil.NewAddressScriptHash(redeemScriptBytes, net)
+	if err != nil {
+		return "", nil
+	}
+
+	return address.String(), nil
+}
+
+func doubleSHA256(payload []byte) ([]byte, error) {
```

Comment:
Why unique function? can we have an unique hash function and re-use it two times where needed?

Reply:
@altafan